File: cpp/memcache.cpp
#include <iostream>
#include <unordered_map>
#include <queue>
#include <functional>
#include <string>
#include <tuple>
#include <functional>
#include <map>
#include <unordered_map>
#include <set>
#include <memory>
using namespace std;

class Memcache {
	map<int, pair<int, int>> data;
public:
	Memcache() {
		// initialize your data structure here.
	}

	int get(int curtTime, int key) {
		// Write your code here
		if (data.find(key) == data.end() || data[key].second < curtTime) {
			return INT_MAX;
		}
		return data[key].first;

	}

	void set(int curtTime, int key, int value, int ttl) {
		// Write your code here
		if (ttl == 0){
			data[key] = make_pair(value, INT_MAX);
		}
		else{
			data[key] = make_pair(value, curtTime + ttl - 1);
		}
		return;
	}

	void del(int curtTime, int key) {
		// Write your code here
		data.erase(key);
		return;
	}

	int incr(int curtTime, int key, int delta) {
		// Write your code here
		if (data.find(key) == data.end() || data[key].second < curtTime) {
			return INT_MAX;
		}
		int newVal = data[key].first + delta;
		data[key] = make_pair(newVal, data[key].second);
		return newVal;

	}

	int decr(int curtTime, int key, int delta) {
		// Write your code here
		if (data.find(key) == data.end() || data[key].second < curtTime) {
			return INT_MAX;
		}
		int newVal = data[key].first - delta;
		data[key] = make_pair(newVal, data[key].second);
		return newVal;
	}
};

int main(){
	shared_ptr<Memcache> memcache = make_shared<Memcache>();
	memcache->set(1, 2, 1, 1);
	cout << memcache->get(2, 2);
	getchar();
	return 0;
}

File: goodyhandy/Memcache.py
class Memcache:

    def __init__(self):
        # initialize your data structure here.
        self.data = {}
        self.INF = 2147483647 
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @return an integer
    def get(self, curtTime, key):
        # Write your code here
        if not self.data.has_key(key) or self.data[key][1] < curtTime:
            return self.INF
        return self.data[key][0]

    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @param {int} value an integer
    # @param {int} ttl an integer
    # @return nothing
    def set(self, curtTime, key, value, ttl):
        # Write your code here
        if ttl == 0:
            self.data[key] = [value, self.INF]
        else:
            self.data[key] = [value, curtTime + ttl - 1]
        return 
        
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @return nothing
    def delete(self, curtTime, key):
        # Write your code here
        del self.data[key]
        return 
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @param {int} delta an integer
    # @return an integer
    def incr(self, curtTime, key, delta):
        # Write your code here
        if not self.data.has_key(key) or self.data[key][1] < curtTime:
            return self.INF
        self.data[key][0] += delta
        return self.data[key][0]

    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @param {int} delta an integer
    # @return an integer
    def decr(self, curtTime, key, delta):
        # Write your code here
        if not self.data.has_key(key) or self.data[key][1] < curtTime:
            return self.INF
        self.data[key][0] -= delta
        return self.data[key][0]

